### PR TITLE
Feature/issue 2418 sgbucket interface

### LINKED ISSUE #2418: Move more functionality into sg-bucket interface and avoid type assertions
## Move GetMaxVbno() into sg-bucket interface 

This code is unnecessarily brittle:

```
	log.Printf("indexWriteBucket: %+v type: %T", k.indexWriteBucket, k.indexWriteBucket)
	cbBucket, ok := k.indexWriteBucket.(base.CouchbaseBucketGoCB)
	var maxVbNo uint16
	if ok {
		maxVbNo, err = cbBucket.GetMaxVbno()
		if err != nil {
			base.Warn("Unable to obtain vbucket count from cluster - defaulting to 1024")
			maxVbNo = 1024
		}
	} else {
		// walrus, for unit testing
		base.Warn("Not a couchbase bucket - assuming walrus, setting maxVbNo=1024")
		maxVbNo = 1024
	}
```

If `k.indexWriteBucket` is a `*base.CouchbaseBucketGoCB` rather than a `base.CouchbaseBucketGoCB`, then this code will break.  So far it looks like that's what's behind https://github.com/couchbaselabs/mobile-testkit/issues/1072

If `GetMaxVbno()` was defined on the sg-bucket bucket interface and all buckets implemented it, it feels like it would be more robust.

### PR DESCRIPTION
**Pull request by tleyden**
Fixes #2418

Dependent PR's:

- https://github.com/couchbase/sg-bucket/pull/20
- https://github.com/couchbaselabs/walrus/pull/27
- https://github.com/couchbaselabs/sync-gateway-accel/pull/133

Functional tests:

- http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-cc/384/ (Passed)
- http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-di/359/ In progress
- http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/277/ Passed
- http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-cc/302/ Passed

### CODE REVIEW

**Comment by tleyden** [comment]
Just realized this is missing an accompanying sg-accel PR, which is why I was seeing [mobile-testkit functional test errors](https://github.com/couchbaselabs/mobile-testkit/issues/1097)

**Comment by tleyden** [comment]
Looks like the sync-gateway-accel functional tests are now failing due to this SGA startup error:

```
2017-04-04T14:58:09.078-04:00 FATAL: Error initializing SG Accel context: Error initializing CBGT: GoCB bucket does not expose UUID -- sg_accel.RunServer() at server.go:48
```

**Comment by tleyden** [comment]
I just saw [this error](https://gist.github.com/tleyden/3f14fa225b6074cb89eec74b0a08bbad) locally:

```
github.com/couchbase/sync_gateway/base.CouchbaseBucketGoCB.ViewCustom(0xc4201b27e0, 0xc4201a1e20, 0x15, 0x4838455, 0x7, 0xc4201ec5f0, 0xc, 0x0, 0x0, 0x0, ...)
	/Users/tleyden/Development/sync_gateway/godeps/src/github.com/couchbase/sync_gateway/base/bucket_gocb.go:1098 +0x51
github.com/couchbase/sync_gateway/base.(*CouchbaseBucketGoCB).ViewCustom(0xc4203b8e00, 0x4839eb7, 0x8, 0x4839e2f, 0x8, 0xc4204e1140, 0x470ee60, 0xc4202bab40, 0x0, 0x0)
	<autogenerated>:408 +0xee
github.com/couchbaselabs/sync-gateway-accel/indexwriter.(*KvChangeIndexWriter).rollbackView(0xc4202a0d00, 0xc42005000f, 0x1, 0x0, 0x4eceea8, 0x100000000, 0xc4200536e8, 0x402bf0c)
	/Users/tleyden/Development/sync_gateway/godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/kv_change_index_writer.go:987 +0x55d
github.com/couchbaselabs/sync-gateway-accel/indexwriter.(*KvChangeIndexWriter).rollbackIndex(0xc4202a0d00, 0xc420053a58, 0xc426553309, 0x4ecf0e0)
	/Users/tleyden/Development/sync_gateway/godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/kv_change_index_writer.go:929 +0x11e
github.com/couchbaselabs/sync-gateway-accel/indexwriter.(*KvChangeIndexWriter).indexEntries(0xc4202a0d00, 0xc4205f4000, 0x1, 0x7d0, 0xc4200e3800, 0x400, 0x400, 0x4c368e0, 0xc420504480, 0x0, ...)
	/Users/tleyden/Development/sync_gateway/godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/kv_change_index_writer.go:638 +0x1348
github.com/couchbaselabs/sync-gateway-accel/indexwriter.(*KvChangeIndexWriter).processPendingQueue(0xc4202a0d00, 0x4900830, 0x4855be9)
	/Users/tleyden/Development/sync_gateway/godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/kv_change_index_writer.go:405 +0x2ab
github.com/couchbaselabs/sync-gateway-accel/indexwriter.(*KvChangeIndexWriter).Init.func1(0xc4202a0d00)
```

and it looks like it's due to this code:

```
func GetGoCBBucketFromBaseBucket(baseBucket Bucket) (bucket CouchbaseBucketGoCB, err error) {
	switch baseBucket := baseBucket.(type) {
	case CouchbaseBucketGoCBGoCouchbaseHybrid:
		return *baseBucket.CouchbaseBucketGoCB, nil
	case *CouchbaseBucketGoCBGoCouchbaseHybrid:
		return *baseBucket.CouchbaseBucketGoCB, nil
	default:
		return CouchbaseBucketGoCB{}, fmt.Errorf("baseBucket %v was not a CouchbaseBucketGoCB.  Was type: %T", baseBucket, baseBucket)
	}
}

```

which is only checking if it's a hybrid bucket, and not just a `CouchbaseBucketGoCB`

File: base/leaky_bucket.go [hunk 1]
```diff
@@ -113,6 +113,10 @@ func (b *LeakyBucket) ViewCustom(ddoc, name string, params map[string]interface{
 	return b.bucket.ViewCustom(ddoc, name, params, vres)
 }
 
+func (b *LeakyBucket) GetMaxVbno() (uint16, error) {
+	return 1024, nil
```
Comment (adamcfraser) [inline base/leaky_bucket.go]: Pass this through to bucket.

File: base/logging_bucket.go [hunk 2]
```diff
@@ -150,3 +150,15 @@ func (b *LoggingBucket) VBHash(docID string) uint32 {
 	LogTo("Bucket", "VBHash()")
 	return b.bucket.VBHash(docID)
 }
+
+func (b *LoggingBucket) GetMaxVbno() (uint16, error) {
+	return 1024, nil
```
Comment (adamcfraser) [inline base/logging_bucket.go]: Pass through.

**Comment by tleyden** [comment]
- [ ] Run sg accel tests

**Comment by tleyden** [comment]
Sg accel tests failing:

```
Running Sync Gateway Accel unit tests
# github.com/couchbaselabs/sync-gateway-accel/indexwriter
godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/dense_channel_storage_test.go:180: cannot use base.NewStatsBucket(indexBucket) (type *base.StatsBucket) as type base.Bucket in assignment:
	*base.StatsBucket does not implement base.Bucket (missing CouchbaseServerVersion method)
godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/dense_channel_storage_test.go:341: impossible type assertion:
	*base.StatsBucket does not implement base.Bucket (missing CouchbaseServerVersion method)
godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/dense_channel_storage_test.go:369: impossible type assertion:
	*base.StatsBucket does not implement base.Bucket (missing CouchbaseServerVersion method)
godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/dense_channel_storage_test.go:398: impossible type assertion:
	*base.StatsBucket does not implement base.Bucket (missing CouchbaseServerVersion method)
godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/dense_channel_storage_test.go:427: impossible type assertion:
	*base.StatsBucket does not implement base.Bucket (missing CouchbaseServerVersion method)
?   	github.com/couchbaselabs/sync-gateway-accel	[no test files]
FAIL	github.com/couchbaselabs/sync-gateway-accel/indexwriter [build failed]
?   	github.com/couchbaselabs/sync-gateway-accel/sg_accel	[no test files]
```

**Comment by tleyden** [comment]
@adamcfraser ok the SGA unit tests are now passing, should be ready to merge

**Pull request merged**